src/dfx/src/commands/canister/update_settings.rs

```diff
@@ -71,51 +73,67 @@ pub async fn exec(
         let mut controllers = controllers;
         let canister_id = CanisterId::from_text(canister_name_or_id)
-            .or_else(|_| canister_id_store.get(canister_name_or_id))?;
+            .or_else(|_| canister_id_store.get(canister_name_or_id))
+            .with_context(|| format!("Failed to get canister id for {}.", canister_name_or_id))?;
         let textual_cid = canister_id.to_text();
         let canister_name = canister_id_store.get_name(&textual_cid).map(|x| &**x);
 
         let compute_allocation = get_compute_allocation(
             opts.compute_allocation.clone(),
             config_interface,
             canister_name,
-        )?;
+        )
+        .context("Failed to get compute allocation.")?;
         let memory_allocation = get_memory_allocation(
             opts.memory_allocation.clone(),
             config_interface,
             canister_name,
-        )?;
+        )
+        .context("Failed to get memory allocation.")?;
         let freezing_threshold = get_freezing_threshold(
             opts.freezing_threshold.clone(),
             config_interface,
             canister_name,
-        )?;
+        )
+        .context("Failed to get freezing threshold.")?;
         if let Some(added) = &opts.add_controller {
-            let status = get_canister_status(env, canister_id, timeout, call_sender).await?;
+            let status = get_canister_status(env, canister_id, timeout, call_sender)
+                .await
+                .with_context(|| format!("Failed to get canister status of {}.", canister_id))?;
             let mut existing_controllers = status.settings.controllers;
             for s in added {
-                existing_controllers.push(controller_to_principal(env, s)?);
+                existing_controllers.push(controller_to_principal(env, s).with_context(|| {
+                    format!("Failed to convert controller {} to principal.", s)
+                })?);
             }
             controllers = Some(existing_controllers);
         }
         if let Some(removed) = &opts.remove_controller {
             let controllers = if opts.add_controller.is_some() {
                 controllers.as_mut().unwrap()
             } else {
-                let status = get_canister_status(env, canister_id, timeout, call_sender).await?;
+                let status = get_canister_status(env, canister_id, timeout, call_sender)
+                    .await
+                    .with_context(|| {
+                        format!("Failed to get canister status for {}.", canister_id)
+                    })?;
                 controllers.get_or_insert(status.settings.controllers)
             };
             let removed = removed
                 .iter()
                 .map(|r| controller_to_principal(env, r))
-                .collect::<DfxResult<Vec<_>>>()?;
+                .collect::<DfxResult<Vec<_>>>()
+                .context("Failed to convert controller to principal.")?;
```

Comment:
This is unlikely to every come up, but I am wondering for cases like this that collect and transpose, would it work to move the context into the mapping?

```
.map(|r| controller_to_principal(env, r)
     .with_context(||format!("Failed to convert controller {} to principal", r)))
.collect:: ...
.context("Failed to convert controllers to principals")
```